File: contains-duplicate-array/java/ContainsDuplicate.java
import java.util.Arrays;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;
import java.util.function.Predicate;

/**
 * Contains Duplicate (LeetCode 217).
 *
 * <p>Given an integer array, return true if any value appears more than once.
 *
 * <p>The four approaches from https://neetcode.io/solutions/contains-duplicate,
 * ordered from brute force to optimal.
 */
public class ContainsDuplicate {

    /**
     * Compares every pair of elements. Time: O(n^2). Space: O(1).
     *
     * <p>The inner loop starts at i + 1, not i, so an element is never compared
     * against itself -- that mistake would report a duplicate for every input.
     */
    static boolean hasDuplicateBruteForce(int[] nums) {
        for (int i = 0; i < nums.length; i++) {
            for (int j = i + 1; j < nums.length; j++) {
                if (nums[i] == nums[j]) {
                    return true;
                }
            }
        }
        return false;
    }

    /**
     * Sorts, then looks for equal neighbours. Time: O(n log n). Space: O(n).
     *
     * <p>Duplicates land next to each other once sorted. Sorts a copy so the
     * caller's array keeps its original order.
     */
    static boolean hasDuplicateSorting(int[] nums) {
        int[] ordered = Arrays.copyOf(nums, nums.length);
        Arrays.sort(ordered);
        for (int i = 1; i < ordered.length; i++) {
            if (ordered[i] == ordered[i - 1]) {
                return true;
            }
        }
        return false;
    }

    /**
     * Remembers values already seen. Time: O(n). Space: O(n).
     *
     * <p>The optimal solution: set lookups are O(1), and it returns on the first
     * repeat instead of always scanning the whole array.
     */
    static boolean hasDuplicateHashSet(int[] nums) {
        Set<Integer> seen = new HashSet<>();
        for (int num : nums) {
            if (seen.contains(num)) {
                return true;
            }
            seen.add(num);
        }
        return false;
    }

    /**
     * Compares the distinct count to the array's length. Time: O(n). Space: O(n).
     *
     * <p>Shortest to write, but it always consumes the whole stream -- no early
     * exit on an array whose first two elements already match.
     */
    static boolean hasDuplicateDistinctCount(int[] nums) {
        return Arrays.stream(nums).distinct().count() < nums.length;
    }

    /** One test case: an expected answer and the input it applies to. */
    private static class Case {
        final boolean expected;
        final int[] nums;

        Case(boolean expected, int... nums) {
            this.expected = expected;
            this.nums = nums;
        }
    }

    public static void main(String[] args) {
        Case[] cases = {
            new Case(true, 1, 2, 3, 3),
            new Case(false, 1, 2, 3, 4),
            new Case(false),
            new Case(false, 7),
            new Case(true, 0, 0),
            new Case(true, -1, -1),
            new Case(false, -1000000000, 1000000000, 0),
        };

        Map<String, Predicate<int[]>> solutions = new LinkedHashMap<>();
        solutions.put("brute force", ContainsDuplicate::hasDuplicateBruteForce);
        solutions.put("sorting", ContainsDuplicate::hasDuplicateSorting);
        solutions.put("hash set", ContainsDuplicate::hasDuplicateHashSet);
        solutions.put("distinct count", ContainsDuplicate::hasDuplicateDistinctCount);

        for (Map.Entry<String, Predicate<int[]>> solution : solutions.entrySet()) {
            int passed = 0;
            for (Case testCase : cases) {
                if (solution.getValue().test(testCase.nums) == testCase.expected) {
                    passed++;
                }
            }
            System.out.printf(
                    "%s  %s: %d/%d cases%n",
                    passed == cases.length ? "PASS" : "FAIL",
                    solution.getKey(),
                    passed,
                    cases.length);
        }

        int[] original = {3, 1, 2};
        for (Predicate<int[]> solve : solutions.values()) {
            solve.test(original);
        }
        boolean unmodified = Arrays.equals(original, new int[] {3, 1, 2});
        System.out.printf(
                "%s  input left unmodified: %s%n",
                unmodified ? "PASS" : "FAIL", Arrays.toString(original));
    }
}

File: contains-duplicate-array/python/contains-duplicate.py
"""Contains Duplicate (LeetCode 217).

Given an integer array, return True if any value appears more than once.

The four approaches from https://neetcode.io/solutions/contains-duplicate,
ordered from brute force to optimal.
"""

from typing import Callable, List, Sequence, Tuple


def has_duplicate_brute_force(nums: List[int]) -> bool:
    """Compare every pair of elements. Time: O(n^2). Space: O(1).

    The inner loop starts at i + 1, not i, so an element is never compared
    against itself -- that mistake would report a duplicate for every input.
    """
    for i in range(len(nums)):
        for j in range(i + 1, len(nums)):
            if nums[i] == nums[j]:
                return True
    return False


def has_duplicate_sorting(nums: List[int]) -> bool:
    """Sort, then look for equal neighbors. Time: O(n log n). Space: O(n).

    Duplicates land next to each other once sorted. Uses sorted() rather than
    nums.sort() so the caller's list keeps its original order.
    """
    ordered = sorted(nums)
    for i in range(1, len(ordered)):
        if ordered[i] == ordered[i - 1]:
            return True
    return False


def has_duplicate_hash_set(nums: List[int]) -> bool:
    """Remember values already seen. Time: O(n). Space: O(n).

    The optimal solution: set lookups are O(1), and it returns on the first
    repeat instead of always scanning the whole array.
    """
    seen = set()
    for num in nums:
        if num in seen:
            return True
        seen.add(num)
    return False


def has_duplicate_set_length(nums: List[int]) -> bool:
    """Compare the set's size to the list's. Time: O(n). Space: O(n).

    A set holds only unique values, so a smaller set means something was
    dropped. Shortest to write, but it always builds the full set -- no early
    exit on an array whose first two elements already match.
    """
    return len(set(nums)) < len(nums)


SOLUTIONS: Tuple[Tuple[str, Callable[[List[int]], bool]], ...] = (
    ("brute force", has_duplicate_brute_force),
    ("sorting", has_duplicate_sorting),
    ("hash set", has_duplicate_hash_set),
    ("set length", has_duplicate_set_length),
)

CASES: Sequence[Tuple[List[int], bool]] = (
    ([1, 2, 3, 3], True),
    ([1, 2, 3, 4], False),
    ([], False),
    ([7], False),
    ([0, 0], True),
    ([-1, -1], True),
    ([-(10**9), 10**9, 0], False),
)


def main() -> None:
    for name, solve in SOLUTIONS:
        passed = sum(solve(list(nums)) == expected for nums, expected in CASES)
        status = "PASS" if passed == len(CASES) else "FAIL"
        print(f"{status}  {name}: {passed}/{len(CASES)} cases")

    original = [3, 1, 2]
    for _, solve in SOLUTIONS:
        solve(original)
    status = "PASS" if original == [3, 1, 2] else "FAIL"
    print(f"{status}  input left unmodified: {original}")


if __name__ == "__main__":
    main()
